File: Python/Distinct Subsequences.py
class Solution:
    # @return an integer
    def numDistinct(self, S, T):
        len1 = len(S)
        len2 = len(T)
        seq = [[0 for col in range(len2+1)] for row in range(len1+1)]
        for i in range(len1+1):
            for j in range(len2+1):
                if i < j:
                    seq [i][j] = 0
                elif i == 0 or j == 0:
                    seq[i][j] = 1
                else:
                    if S[i-1] != T[j-1]:
                        seq[i][j] = seq[i-1][j];
                    else:
                        seq[i][j] = seq[i-1][j] + seq[i-1][j-1]
        return seq[len1][len2]

File: C++/Distinct-Subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        if (t.empty() || s.empty()) return 0;
        int lens = s.size(), lent = t.size();
        vector<vector<int> > dp(lens + 1, vector<int>(lent + 1, 0));
        for (int i = 0; i <= lens; i ++) {
            for (int j = 0; j <= lent; j ++) {
                if (i < j) dp[i][j] = 0;
                else if (i == 0 || j == 0) dp[i][j] = 1;
                else if(s[i - 1] == t[j - 1]) dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j];
                else dp[i][j] = dp[i - 1][j];
            }
        }
        return dp[lens][lent];
    }
};